Enable content clipping for LineEdit, add TextServer option to clamp glyph offsets.

bruvzg opened the pull request:
Fixes https://github.com/godotengine/godot/issues/87016
Fixes https://github.com/godotengine/godot-proposals/issues/8860

- Enable content clipping for `LineEdit` (via separate canvas item for the text).
- Add `TextServer` option to clamp glyph offsets (global default, editor setting, and API for overriding it in the custom drawing).

| Before | After (no clamping) | After (with clamping set to `1.0`) |
| --- | --- | --- |
| <img width="419" alt="Screenshot 2024-01-10 at 10 42 53" src="https://github.com/godotengine/godot/assets/7645683/a8e202c1-117f-4940-8c76-713ab5dcef39"> | <img width="419" alt="Screenshot 2024-01-10 at 11 16 59" src="https://github.com/godotengine/godot/assets/7645683/ad76a0ef-2e84-49a0-9d35-8de8b54ead51"> | <img width="419" alt="Screenshot 2024-01-10 at 11 34 11" src="https://github.com/godotengine/godot/assets/7645683/166b82cf-a568-478e-8870-bbabac4da4e7"> |

bruvzg committed:
Enable content clipping for LineEdit, add TextServer option to clamp glyph offsets.

Calinou commented on doc/classes/EditorSettings.xml:
```diff
@@ -576,6 +576,9 @@
 			FreeType's font anti-aliasing mode used to render the editor fonts. Most fonts are not designed to look good with anti-aliasing disabled, so it's recommended to leave this enabled unless you're using a pixel art font.
 		</member>
+		<member name="interface/editor/font_clamp_glyph_offset" type="float" setter="" getter="">
+			Sets default value for glyph offset clamping. Value is a fraction of font height. If set to zero or negative value, no clamping is done.
```
Out of curiosity, why does this need to be an editor setting (as opposed to say, a project setting)?

bruvzg replied:
I guess project setting also should be added.

bruvzg replied:
Moved to project settings.

Calinou commented on scene/gui/line_edit.cpp:
```diff
@@ -810,6 +810,8 @@ void LineEdit::_notification(int p_what) {
 
 			RID ci = get_canvas_item();
 
+			RenderingServer::get_singleton()->canvas_item_set_clip(ci, true);
```
Note that this may negatively affect focus StyleBox drawing, in a manner similar to TextEdit: https://github.com/godotengine/godot/issues/32996

Kiisu-Master replied:
Not only the focus stylebox but also children of the node.

lostminds commented:
After looking at the code a little closer I see now that LineEdit does its own rendering of text, would it not be possible to have the LineEdit maintain an internal RichTextLabel child node for this purpose? A regular Label would probably be enough in most cases, but a RichTextLabel would allow rendering separate selection color etc. This label node could then be clipped separately, solving this issue and https://github.com/godotengine/godot-proposals/issues/8860 without needing to make the entire LineEdit control clip children with the issues that would entail as pointed out above? It could also reduce some code duplication as the text rendering code would no longer be needed in LineEdit.

bruvzg commented:
Adding a full Label is overkill, but adding a separate canvas item to draw should probably work.

bruvzg commented:
Added a separate canvas item for the text, now it should clip only text. Also changed clipping check to implement https://github.com/godotengine/godot-proposals/issues/8860 as well.

dhoverml commented:
I tried cherry-picking this to master (f317cc713a), but got two merge conflicts in `modules/text_server_adv/text_server_adv.cpp` (I believe with 85df221610e)
